src/Common/TimerService.h

```diff
@@ -26,6 +28,7 @@ class TimerService final
     /// Safe to call from other threads.
     muduo::net::TimerId runAfter(double delay, muduo::net::TimerCallback cb)
     {
+        std::scoped_lock lock(event_loop_mutex);
         assert(event_loop);
         return event_loop->runAfter(delay, std::move(cb));
```

Comment:
After TimerService shutdown, the event_loop is nullptr, but it still may be called in `CheckpointCoordinator::removeExpiredCheckpoints` ?

Reply:
We can make the class `TimerService` is thread safe, or we can also ensure it where used

Reply:
that's the reason I added a lock guard here, and checked the pointer exists in the shutdown function.
Do you mean we need to check the pointer exists before calling the function in the `runAfter` ? 
```
void TimerService::shutdown()
{
    {
        std::scoped_lock lock(event_loop_mutex);
        if (event_loop)
        {
            event_loop->quit();

            /// Decrement ref count to make sure the last ref count
            /// of event loop is in the creation thread
            event_loop = nullptr;
        }
    }

    eloop_init_guard.store(nullptr);
    eloop_init_guard.notify_all();

    looper.wait();
}
```

Reply:
yes, as above mentioned that when quit proton-server muduo timeservice data race https://github.com/timeplus-io/proton/issues/320#issuecomment-1821966387 ,
One thread is calling  `TimerService::shutdown()` in `~CheckpointCoordinator()`
Another thread is calling `TimerService::runAfter()` (belongs to EventLoop thread)